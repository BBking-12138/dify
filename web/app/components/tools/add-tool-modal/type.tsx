'use client'
import cn from 'classnames'
import { useTranslation } from 'react-i18next'

type Props = {
  value: string
  onSelect: (type: string) => void
}

const Types = ({
  value,
  onSelect,
}: Props) => {
  const { t } = useTranslation()

  return (
    <div className='mb-3'>
      <div className='px-3 py-0.5 text-gray-500 text-xs leading-[18px] font-medium'>{t('tools.addToolModal.type').toLocaleUpperCase()}</div>
      <div className={cn('mb-0.5 p-1 pl-3 flex items-center cursor-pointer text-gray-700 text-sm leading-5 rounded-lg hover:bg-white', value === 'all' && '!bg-white !text-primary-600 font-medium')} onClick={() => onSelect('all')}>{t('tools.type.all')}</div>
      <div className={cn('mb-0.5 p-1 pl-3 flex items-center cursor-pointer text-gray-700 text-sm leading-5 rounded-lg hover:bg-white', value === 'builtin' && '!bg-white !text-primary-600 font-medium')} onClick={() => onSelect('builtin')}>{t('tools.type.builtIn')}</div>
      <div className={cn('mb-0.5 p-1 pl-3 flex items-center cursor-pointer text-gray-700 text-sm leading-5 rounded-lg hover:bg-white', value === 'custom' && '!bg-white !text-primary-600 font-medium')} onClick={() => onSelect('api')}>{t('tools.type.custom')}</div>
      <div className={cn('mb-0.5 p-1 pl-3 flex items-center cursor-pointer text-gray-700 text-sm leading-5 rounded-lg hover:bg-white', value === 'workflow' && '!bg-white !text-primary-600 font-medium')} onClick={() => onSelect('workflow')}>{t('tools.type.workflow')}</div>
    </div>
  )
}
export default Types
