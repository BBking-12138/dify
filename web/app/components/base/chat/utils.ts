import { addFileInfos, sortAgentSorts } from '../../tools/utils'
import { UUID_NIL } from './constants'
import type { IChatItem } from './chat/type'
import type { ChatItem, ChatItemInTree } from './types'

async function decodeBase64AndDecompress(base64String: string) {
  const binaryString = atob(base64String)
  const compressedUint8Array = Uint8Array.from(binaryString, char => char.charCodeAt(0))
  const decompressedStream = new Response(compressedUint8Array).body?.pipeThrough(new DecompressionStream('gzip'))
  const decompressedArrayBuffer = await new Response(decompressedStream).arrayBuffer()
  return new TextDecoder().decode(decompressedArrayBuffer)
}

function getProcessedInputsFromUrlParams(): Record<string, any> {
  const urlParams = new URLSearchParams(window.location.search)
  const inputs: Record<string, any> = {}
  urlParams.forEach(async (value, key) => {
    inputs[key] = await decodeBase64AndDecompress(decodeURIComponent(value))
  })
  return inputs
}

function appendQAToChatList(chatList: ChatItem[], item: any) {
  // we append answer first and then question since will reverse the whole chatList later
  chatList.push({
    id: item.id,
    content: item.answer,
    agent_thoughts: addFileInfos(item.agent_thoughts ? sortAgentSorts(item.agent_thoughts) : item.agent_thoughts, item.message_files),
    feedback: item.feedback,
    isAnswer: true,
    citation: item.retriever_resources,
    message_files: item.message_files?.filter((file: any) => file.belongs_to === 'assistant') || [],
  })
  chatList.push({
    id: `question-${item.id}`,
    content: item.query,
    isAnswer: false,
    message_files: item.message_files?.filter((file: any) => file.belongs_to === 'user') || [],
  })
}

/**
 * Computes the latest thread messages from all messages of the conversation.
 * Same logic as backend codebase `api/core/prompt/utils/extract_thread_messages.py`
 *
 * @param fetchedMessages - The history chat list data from the backend, sorted by created_at in descending order. This includes all flattened history messages of the conversation.
 * @returns An array of ChatItems representing the latest thread.
 */
function getPrevChatList(fetchedMessages: any[]) {
  const ret: ChatItem[] = []
  let nextMessageId = null

  for (const item of fetchedMessages) {
    if (!item.parent_message_id) {
      appendQAToChatList(ret, item)
      break
    }

    if (!nextMessageId) {
      appendQAToChatList(ret, item)
      nextMessageId = item.parent_message_id
    }
    else {
      if (item.id === nextMessageId || nextMessageId === UUID_NIL) {
        appendQAToChatList(ret, item)
        nextMessageId = item.parent_message_id
      }
    }
  }
  return ret.reverse()
}

function buildChatItemTree(allMessages: IChatItem[]): ChatItemInTree[] {
  const map: Record<string, ChatItemInTree> = {}
  let rootNodes: ChatItemInTree[] = []
  const childrenCount: Record<string, number> = {}

  const legacyQuestions: ChatItemInTree[] = []

  for (let i = 0; i < allMessages.length; i += 2) {
    const question = allMessages[i]!
    const answer = allMessages[i + 1]!

    // Process question
    const parentId = question.parentMessageId ?? ''
    childrenCount[parentId] = (childrenCount[parentId] || 0) + 1
    const questionNode: ChatItemInTree = {
      ...question,
      children: [],
      siblingIndex: childrenCount[parentId] - 1,
    }
    map[question.id] = questionNode

    // Process answer
    childrenCount[question.id] = 1
    const answerNode: ChatItemInTree = {
      ...answer,
      children: [],
      siblingIndex: 0,
    }
    map[answer.id] = answerNode

    // Connect question and answer
    questionNode.children!.push(answerNode)

    // Connect to parent or add to root
    if (!parentId)
      rootNodes.push(questionNode)
    else if (parentId !== UUID_NIL)
      map[parentId]?.children!.push(questionNode)
    else
      legacyQuestions.unshift(questionNode)
  }

  // legacy message compat
  for (const legacyQuestion of legacyQuestions) {
    const answer = legacyQuestion.children![0]!
    const questionNode = map[legacyQuestion.id]
    const answerNode = map[answer.id]
    answerNode.children?.unshift(...rootNodes)
    rootNodes = [questionNode]
  }

  return rootNodes
}

function getThreadMessages(tree: ChatItemInTree[], targetMessageId?: string): ChatItemInTree[] {
  let ret: ChatItemInTree[] = []
  let targetNode: ChatItemInTree | undefined

  // find path to the target message
  const stack = tree.map(rootNode => ({
    node: rootNode,
    path: [rootNode],
  }))
  while (stack.length > 0) {
    const { node, path } = stack.pop()!
    if (node.id === targetMessageId) {
      targetNode = node
      ret = path
      break
    }
    if (node.children) {
      stack.push(...node.children.map(child => ({
        node: child,
        path: [...path, child],
      })))
    }
  }

  // append all descendant messages to the path
  if (targetNode) {
    const stack = [targetNode]
    while (stack.length > 0) {
      const node = stack.pop()!
      if (node !== targetNode)
        ret.push(node)
      if (node.children?.length)
        stack.push(node.children.at(-1)!)
    }
  }
  return ret
}

export {
  getProcessedInputsFromUrlParams,
  getPrevChatList,
  buildChatItemTree,
  getThreadMessages,
}
