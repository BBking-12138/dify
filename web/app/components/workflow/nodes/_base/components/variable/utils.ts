import produce from 'immer'
import { isArray, uniq } from 'lodash-es'
import type { CodeNodeType } from '../../../code/types'
import type { EndNodeType } from '../../../end/types'
import type { AnswerNodeType } from '../../../answer/types'
import type { LLMNodeType } from '../../../llm/types'
import type { KnowledgeRetrievalNodeType } from '../../../knowledge-retrieval/types'
import type { IfElseNodeType } from '../../../if-else/types'
import type { TemplateTransformNodeType } from '../../../template-transform/types'
import type { QuestionClassifierNodeType } from '../../../question-classifier/types'
import type { HttpNodeType } from '../../../http/types'
import { VarType as ToolVarType } from '../../../tool/types'
import type { ToolNodeType } from '../../../tool/types'
import { BlockEnum, InputVarType, VarType } from '@/app/components/workflow/types'
import type { StartNodeType } from '@/app/components/workflow/nodes/start/types'
import type { Node, NodeOutPutVar, ValueSelector, Var } from '@/app/components/workflow/types'
import type { VariableAssignerNodeType } from '@/app/components/workflow/nodes/variable-assigner/types'
import {
  CHAT_QUESTION_CLASSIFIER_OUTPUT_STRUCT,
  COMPLETION_QUESTION_CLASSIFIER_OUTPUT_STRUCT,
  HTTP_REQUEST_OUTPUT_STRUCT,
  KNOWLEDGE_RETRIEVAL_OUTPUT_STRUCT,
  LLM_OUTPUT_STRUCT,
  SUPPORT_OUTPUT_VARS_NODE,
  TEMPLATE_TRANSFORM_OUTPUT_STRUCT,
  TOOL_OUTPUT_STRUCT,
} from '@/app/components/workflow/constants'
import type { PromptItem } from '@/models/debug'
import { VAR_REGEX } from '@/config'

const inputVarTypeToVarType = (type: InputVarType): VarType => {
  if (type === InputVarType.number)
    return VarType.number

  return VarType.string
}

const findExceptVarInObject = (obj: any, filterVar: (payload: Var, selector: ValueSelector) => boolean, value_selector: ValueSelector): Var => {
  const { children } = obj
  const res: Var = {
    variable: obj.variable,
    type: VarType.object,
    children: children.filter((item: Var) => {
      const { children } = item
      const currSelector = [...value_selector, item.variable]
      if (!children)
        return filterVar(item, currSelector)

      const obj = findExceptVarInObject(item, filterVar, currSelector)
      return obj.children && obj.children?.length > 0
    }),
  }
  return res
}

const formatItem = (item: any, isChatMode: boolean, filterVar: (payload: Var, selector: ValueSelector) => boolean): NodeOutPutVar => {
  const { id, data } = item
  const res: NodeOutPutVar = {
    nodeId: id,
    title: data.title,
    vars: [],
  }
  switch (data.type) {
    case BlockEnum.Start: {
      const {
        variables,
      } = data as StartNodeType
      res.vars = variables.map((v) => {
        return {
          variable: v.variable,
          type: inputVarTypeToVarType(v.type),
          isParagraph: v.type === InputVarType.paragraph,
          isSelect: v.type === InputVarType.select,
          options: v.options,
          required: v.required,
        }
      })
      if (isChatMode) {
        res.vars.push({
          variable: 'sys.query',
          type: VarType.string,
        })
        res.vars.push({
          variable: 'sys.conversation_id',
          type: VarType.string,
        })
      }
      res.vars.push({
        variable: 'sys.user_id',
        type: VarType.string,
      })
      res.vars.push({
        variable: 'sys.files',
        type: VarType.arrayFile,
      })
      break
    }

    case BlockEnum.LLM: {
      res.vars = LLM_OUTPUT_STRUCT
      break
    }

    case BlockEnum.KnowledgeRetrieval: {
      res.vars = KNOWLEDGE_RETRIEVAL_OUTPUT_STRUCT
      break
    }

    case BlockEnum.Code: {
      const {
        outputs,
      } = data as CodeNodeType
      res.vars = Object.keys(outputs).map((key) => {
        return {
          variable: key,
          type: outputs[key].type,
        }
      })
      break
    }

    case BlockEnum.TemplateTransform: {
      res.vars = TEMPLATE_TRANSFORM_OUTPUT_STRUCT
      break
    }

    case BlockEnum.QuestionClassifier: {
      res.vars = isChatMode ? CHAT_QUESTION_CLASSIFIER_OUTPUT_STRUCT : COMPLETION_QUESTION_CLASSIFIER_OUTPUT_STRUCT
      break
    }

    case BlockEnum.HttpRequest: {
      res.vars = HTTP_REQUEST_OUTPUT_STRUCT
      break
    }

    case BlockEnum.VariableAssigner: {
      const {
        output_type,
      } = data as VariableAssignerNodeType
      res.vars = [
        {
          variable: 'output',
          type: output_type,
        },
      ]
      break
    }

    case BlockEnum.Tool: {
      res.vars = TOOL_OUTPUT_STRUCT
      break
    }
  }

  const selector = [id]
  res.vars = res.vars.filter((v) => {
    const { children } = v
    if (!children)
      return filterVar(v, selector)

    const obj = findExceptVarInObject(v, filterVar, selector)
    return obj?.children && obj?.children.length > 0
  }).map((v) => {
    const { children } = v
    if (!children)
      return v

    return findExceptVarInObject(v, filterVar, selector)
  })

  return res
}
export const toNodeOutputVars = (nodes: any[], isChatMode: boolean, filterVar = (_payload: Var, _selector: ValueSelector) => true): NodeOutPutVar[] => {
  const res = nodes
    .filter(node => SUPPORT_OUTPUT_VARS_NODE.includes(node.data.type))
    .map((node) => {
      return {
        ...formatItem(node, isChatMode, filterVar),
        isStartNode: node.data.type === BlockEnum.Start,
      }
    })
    .filter(item => item.vars.length > 0)
  return res
}

export const isSystemVar = (valueSelector: ValueSelector) => {
  return valueSelector[0] === 'sys' || valueSelector[1] === 'sys'
}

export const getNodeInfoById = (nodes: any, id: string) => {
  if (!isArray(nodes))
    return
  return nodes.find((node: any) => node.id === id)
}

export const getVarType = (value: ValueSelector, availableNodes: any[], isChatMode: boolean): VarType | undefined => {
  const isSystem = isSystemVar(value)
  const startNode = availableNodes.find((node: any) => {
    return node.data.type === BlockEnum.Start
  })
  const allOutputVars = toNodeOutputVars(availableNodes, isChatMode)

  const targetVarNodeId = isSystem ? startNode?.id : value[0]
  const targetVar = allOutputVars.find(v => v.nodeId === targetVarNodeId)

  if (!targetVar)
    return undefined

  let type: VarType = VarType.string
  let curr: any = targetVar.vars
  if (isSystem) {
    return curr.find((v: any) => v.variable === (value as ValueSelector).join('.'))?.type
  }
  else {
    (value as ValueSelector).slice(1).forEach((key, i) => {
      const isLast = i === value.length - 2
      curr = curr.find((v: any) => v.variable === key)
      if (isLast) {
        type = curr?.type
      }
      else {
        if (curr.type === VarType.object)
          curr = curr.children
      }
    })
    return type
  }
}

const matchNotSystemVars = (prompts: string[]) => {
  if (!prompts)
    return []

  const allVars: string[] = []
  prompts.forEach((prompt) => {
    VAR_REGEX.lastIndex = 0
    allVars.push(...(prompt.match(VAR_REGEX) || []))
  })
  const uniqVars = uniq(allVars).map(v => v.replaceAll('{{#', '').replace('#}}', '').split('.'))
  return uniqVars
}

const replaceOldVarInText = (text: string, oldVar: ValueSelector, newVar: ValueSelector) => {
  if (!text || typeof text !== 'string')
    return text

  if (!newVar || newVar.length === 0)
    return text

  return text.replaceAll(`{{#${oldVar.join('.')}#}}`, `{{#${newVar.join('.')}#}}`)
}

export const getNodeUsedVars = (node: Node): ValueSelector[] => {
  const { data } = node
  const { type } = data
  let res: ValueSelector[] = []
  switch (type) {
    case BlockEnum.End: {
      res = (data as EndNodeType).outputs?.map((output) => {
        return output.value_selector
      })
      break
    }
    case BlockEnum.Answer: {
      res = (data as AnswerNodeType).variables?.map((v) => {
        return v.value_selector
      })
      break
    }
    case BlockEnum.LLM: {
      const payload = (data as LLMNodeType)
      const isChatModel = payload.model?.mode === 'chat'
      let prompts: string[] = []
      if (isChatModel)
        prompts = (payload.prompt_template as PromptItem[])?.map(p => p.text) || []
      else
        prompts = [(payload.prompt_template as PromptItem).text]

      const inputVars: ValueSelector[] = matchNotSystemVars(prompts)
      const contextVar = (data as LLMNodeType).context?.variable_selector ? [(data as LLMNodeType).context?.variable_selector] : []
      res = [...inputVars, ...contextVar]
      break
    }
    case BlockEnum.KnowledgeRetrieval: {
      res = [(data as KnowledgeRetrievalNodeType).query_variable_selector]
      break
    }
    case BlockEnum.IfElse: {
      res = (data as IfElseNodeType).conditions?.map((c) => {
        return c.variable_selector
      })
      break
    }
    case BlockEnum.Code: {
      res = (data as CodeNodeType).variables?.map((v) => {
        return v.value_selector
      })
      break
    }
    case BlockEnum.TemplateTransform: {
      res = (data as TemplateTransformNodeType).variables?.map((v: any) => {
        return v.value_selector
      })
      break
    }
    case BlockEnum.QuestionClassifier: {
      res = [(data as QuestionClassifierNodeType).query_variable_selector]
      break
    }
    case BlockEnum.HttpRequest: {
      const payload = (data as HttpNodeType)
      res = matchNotSystemVars([payload.url, payload.headers, payload.params, payload.body.data])
      break
    }
    case BlockEnum.Tool: {
      const payload = (data as ToolNodeType)
      const mixVars = matchNotSystemVars(Object.keys(payload.tool_parameters)?.filter(key => payload.tool_parameters[key].type === ToolVarType.mixed).map(key => payload.tool_parameters[key].value) as string[])
      const vars = Object.keys(payload.tool_parameters).filter(key => payload.tool_parameters[key].type === ToolVarType.variable).map(key => payload.tool_parameters[key].value as string) || []
      res = [...(mixVars as ValueSelector[]), ...(vars as any)]
      break
    }

    case BlockEnum.VariableAssigner: {
      res = (data as VariableAssignerNodeType)?.variables
    }
  }
  return res || []
}

export const findUsedVarNodes = (varSelector: ValueSelector, availableNodes: Node[]): Node[] => {
  const res: Node[] = []
  availableNodes.forEach((node) => {
    const vars = getNodeUsedVars(node)
    if (vars.find(v => v.join('.') === varSelector.join('.')))
      res.push(node)
  })
  return res
}

export const updateNodeVars = (oldNode: Node, oldVarSelector: ValueSelector, newVarSelector: ValueSelector): Node => {
  const newNode = produce(oldNode, (draft: any) => {
    const { data } = draft
    const { type } = data
    switch (type) {
      case BlockEnum.End: {
        const payload = data as EndNodeType
        if (payload.outputs) {
          payload.outputs = payload.outputs.map((output) => {
            if (output.value_selector.join('.') === oldVarSelector.join('.'))
              output.value_selector = newVarSelector
            return output
          })
        }
        break
      }
      case BlockEnum.Answer: {
        const payload = data as AnswerNodeType
        if (payload.variables) {
          payload.variables = payload.variables.map((v) => {
            if (v.value_selector.join('.') === oldVarSelector.join('.'))
              v.value_selector = newVarSelector
            return v
          })
        }
        break
      }
      case BlockEnum.LLM: {
        const payload = data as LLMNodeType
        const isChatModel = payload.model?.mode === 'chat'
        if (isChatModel) {
          payload.prompt_template = (payload.prompt_template as PromptItem[]).map((prompt) => {
            return {
              ...prompt,
              text: replaceOldVarInText(prompt.text, oldVarSelector, newVarSelector),
            }
          })
        }
        else {
          payload.prompt_template = {
            ...payload.prompt_template,
            text: replaceOldVarInText((payload.prompt_template as PromptItem).text, oldVarSelector, newVarSelector),
          }
        }
        if (payload.context?.variable_selector?.join('.') === oldVarSelector.join('.'))
          payload.context.variable_selector = newVarSelector

        break
      }
      case BlockEnum.KnowledgeRetrieval: {
        const payload = data as KnowledgeRetrievalNodeType
        if (payload.query_variable_selector.join('.') === oldVarSelector.join('.'))
          payload.query_variable_selector = newVarSelector
        break
      }
      case BlockEnum.IfElse: {
        const payload = data as IfElseNodeType
        if (payload.conditions) {
          payload.conditions = payload.conditions.map((c) => {
            if (c.variable_selector.join('.') === oldVarSelector.join('.'))
              c.variable_selector = newVarSelector
            return c
          })
        }
        break
      }
      case BlockEnum.Code: {
        const payload = data as CodeNodeType
        if (payload.variables) {
          payload.variables = payload.variables.map((v) => {
            if (v.value_selector.join('.') === oldVarSelector.join('.'))
              v.value_selector = newVarSelector
            return v
          })
        }
        break
      }
      case BlockEnum.TemplateTransform: {
        const payload = data as TemplateTransformNodeType
        if (payload.variables) {
          payload.variables = payload.variables.map((v: any) => {
            if (v.value_selector.join('.') === oldVarSelector.join('.'))
              v.value_selector = newVarSelector
            return v
          })
        }
        break
      }
      case BlockEnum.QuestionClassifier: {
        const payload = data as QuestionClassifierNodeType
        if (payload.query_variable_selector.join('.') === oldVarSelector.join('.'))
          payload.query_variable_selector = newVarSelector
        break
      }
      case BlockEnum.HttpRequest: {
        const payload = data as HttpNodeType
        payload.url = replaceOldVarInText(payload.url, oldVarSelector, newVarSelector)
        payload.headers = replaceOldVarInText(payload.headers, oldVarSelector, newVarSelector)
        payload.params = replaceOldVarInText(payload.params, oldVarSelector, newVarSelector)
        payload.body.data = replaceOldVarInText(payload.body.data, oldVarSelector, newVarSelector)
        break
      }
      case BlockEnum.Tool: {
        const payload = data as ToolNodeType
        const hasShouldRenameVar = Object.keys(payload.tool_parameters)?.filter(key => payload.tool_parameters[key].type !== ToolVarType.constant)
        if (hasShouldRenameVar) {
          Object.keys(payload.tool_parameters).forEach((key) => {
            const value = payload.tool_parameters[key]
            const { type } = value
            if (type === ToolVarType.variable) {
              payload.tool_parameters[key] = {
                ...value,
                value: newVarSelector,
              }
            }

            if (type === ToolVarType.mixed) {
              payload.tool_parameters[key] = {
                ...value,
                value: replaceOldVarInText(payload.tool_parameters[key].value as string, oldVarSelector, newVarSelector),
              }
            }
          })
        }
        break
      }
      case BlockEnum.VariableAssigner: {
        const payload = data as VariableAssignerNodeType
        if (payload.variables) {
          payload.variables = payload.variables.map((v) => {
            if (v.join('.') === oldVarSelector.join('.'))
              v = newVarSelector
            return v
          })
        }
        break
      }
    }
  })
  return newNode
}
const varToValueSelectorList = (v: Var, parentValueSelector: ValueSelector, res: ValueSelector[]) => {
  if (!v.variable)
    return

  res.push([...parentValueSelector, v.variable])

  if (v.children && v.children.length > 0) {
    v.children.forEach((child) => {
      varToValueSelectorList(child, [...parentValueSelector, v.variable], res)
    })
  }
}

const varsToValueSelectorList = (vars: Var | Var[], parentValueSelector: ValueSelector, res: ValueSelector[]) => {
  if (Array.isArray(vars)) {
    vars.forEach((v) => {
      varToValueSelectorList(v, parentValueSelector, res)
    })
  }
  varToValueSelectorList(vars as Var, parentValueSelector, res)
}

export const getNodeOutputVars = (node: Node, isChatMode: boolean): ValueSelector[] => {
  const { data, id } = node
  const { type } = data
  let res: ValueSelector[] = []

  switch (type) {
    case BlockEnum.Start: {
      const {
        variables,
      } = data as StartNodeType
      res = variables.map((v) => {
        return [id, v.variable]
      })

      if (isChatMode) {
        res.push([id, 'sys', 'query'])
        res.push([id, 'sys', 'files'])
      }
      break
    }

    case BlockEnum.LLM: {
      varsToValueSelectorList(LLM_OUTPUT_STRUCT, [id], res)
      break
    }

    case BlockEnum.KnowledgeRetrieval: {
      varsToValueSelectorList(KNOWLEDGE_RETRIEVAL_OUTPUT_STRUCT, [id], res)
      break
    }

    case BlockEnum.Code: {
      const {
        outputs,
      } = data as CodeNodeType
      Object.keys(outputs).forEach((key) => {
        res.push([id, key])
      })
      break
    }

    case BlockEnum.TemplateTransform: {
      varsToValueSelectorList(TEMPLATE_TRANSFORM_OUTPUT_STRUCT, [id], res)
      break
    }

    case BlockEnum.QuestionClassifier: {
      varsToValueSelectorList(isChatMode ? CHAT_QUESTION_CLASSIFIER_OUTPUT_STRUCT : COMPLETION_QUESTION_CLASSIFIER_OUTPUT_STRUCT, [id], res)
      break
    }

    case BlockEnum.HttpRequest: {
      varsToValueSelectorList(HTTP_REQUEST_OUTPUT_STRUCT, [id], res)
      break
    }

    case BlockEnum.VariableAssigner: {
      res.push([id, 'output'])
      break
    }

    case BlockEnum.Tool: {
      varsToValueSelectorList(TOOL_OUTPUT_STRUCT, [id], res)
      break
    }
  }

  return res
}
