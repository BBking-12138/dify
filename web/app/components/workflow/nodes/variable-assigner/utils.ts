import type { VariableAssignerNodeType } from './types'

export const checkNodeValid = (payload: VariableAssignerNodeType) => {
  return true
}
