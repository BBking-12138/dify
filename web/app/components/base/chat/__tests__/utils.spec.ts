import type { IChatItem } from '../chat/type'
import { buildChatItemTree } from '../utils'
import branchedTestMessages from './branchedTestMessages.json'
import legacyTestMessages from './legacyTestMessages.json'
import mixedTestMessages from './mixedTestMessages.json'
import multiRootNodesMessages from './multiRootNodesMessages.json'
import multiRootNodesWithLegacyTestMessages from './multiRootNodesWithLegacyTestMessages.json'

describe('buildChatItemTree', () => {
  it('should build a tree from a list of chat items', () => {
    const tree1 = buildChatItemTree(branchedTestMessages as IChatItem[])
    expect(tree1).toMatchSnapshot()
  })

  it('should be compatible with legacy chat items', () => {
    const tree2 = buildChatItemTree(legacyTestMessages as IChatItem[])
    expect(tree2).toMatchSnapshot()
  })

  it('should build a tree from a list of mixed chat items', () => {
    const tree3 = buildChatItemTree(mixedTestMessages as IChatItem[])
    expect(tree3).toMatchSnapshot()
  })

  it('should build a tree from a list of multi root nodes chat items', () => {
    const tree4 = buildChatItemTree(multiRootNodesMessages as IChatItem[])
    expect(tree4).toMatchSnapshot()
  })

  it('should build a tree from a list of multi root nodes chat items with legacy chat items', () => {
    const tree5 = buildChatItemTree(multiRootNodesWithLegacyTestMessages as IChatItem[])
    expect(tree5).toMatchSnapshot()
  })
})
