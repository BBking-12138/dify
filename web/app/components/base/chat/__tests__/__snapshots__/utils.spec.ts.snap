// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildChatItemTree should be compatible with legacy chat items 1`] = `
[
  {
    "children": [
      {
        "children": [
          {
            "children": [
              {
                "children": [
                  {
                    "children": [
                      {
                        "children": [
                          {
                            "children": [
                              {
                                "children": [],
                                "id": "4",
                                "isAnswer": true,
                                "parentMessageId": "question-4",
                                "siblingIndex": 0,
                              },
                            ],
                            "id": "question-4",
                            "isAnswer": false,
                            "parentMessageId": "00000000-0000-0000-0000-000000000000",
                            "siblingIndex": 0,
                          },
                        ],
                        "id": "3",
                        "isAnswer": true,
                        "parentMessageId": "question-3",
                        "siblingIndex": 0,
                      },
                    ],
                    "id": "question-3",
                    "isAnswer": false,
                    "parentMessageId": "00000000-0000-0000-0000-000000000000",
                    "siblingIndex": 0,
                  },
                ],
                "id": "2",
                "isAnswer": true,
                "parentMessageId": "question-2",
                "siblingIndex": 0,
              },
            ],
            "id": "question-2",
            "isAnswer": false,
            "parentMessageId": "00000000-0000-0000-0000-000000000000",
            "siblingIndex": 0,
          },
        ],
        "id": "1",
        "isAnswer": true,
        "parentMessageId": "question-1",
        "siblingIndex": 0,
      },
    ],
    "id": "question-1",
    "isAnswer": false,
    "parentMessageId": "00000000-0000-0000-0000-000000000000",
    "siblingIndex": 0,
  },
]
`;

exports[`buildChatItemTree should build a tree from a list of chat items 1`] = `
[
  {
    "children": [
      {
        "children": [
          {
            "children": [
              {
                "children": [
                  {
                    "children": [
                      {
                        "children": [],
                        "id": "3",
                        "isAnswer": true,
                        "parentMessageId": "question-3",
                        "siblingIndex": 0,
                      },
                    ],
                    "id": "question-3",
                    "isAnswer": false,
                    "parentMessageId": "2",
                    "siblingIndex": 0,
                  },
                ],
                "id": "2",
                "isAnswer": true,
                "parentMessageId": "question-2",
                "siblingIndex": 0,
              },
            ],
            "id": "question-2",
            "isAnswer": false,
            "parentMessageId": "1",
            "siblingIndex": 0,
          },
          {
            "children": [
              {
                "children": [],
                "id": "4",
                "isAnswer": true,
                "parentMessageId": "question-4",
                "siblingIndex": 0,
              },
            ],
            "id": "question-4",
            "isAnswer": false,
            "parentMessageId": "1",
            "siblingIndex": 1,
          },
        ],
        "id": "1",
        "isAnswer": true,
        "parentMessageId": "question-1",
        "siblingIndex": 0,
      },
    ],
    "id": "question-1",
    "isAnswer": false,
    "parentMessageId": null,
    "siblingIndex": 0,
  },
]
`;

exports[`buildChatItemTree should build a tree from a list of mixed chat items 1`] = `
[
  {
    "children": [
      {
        "children": [
          {
            "children": [
              {
                "children": [
                  {
                    "children": [
                      {
                        "children": [],
                        "id": "3",
                        "isAnswer": true,
                        "parentMessageId": "question-3",
                        "siblingIndex": 0,
                      },
                    ],
                    "id": "question-3",
                    "isAnswer": false,
                    "parentMessageId": "2",
                    "siblingIndex": 0,
                  },
                ],
                "id": "2",
                "isAnswer": true,
                "parentMessageId": "question-2",
                "siblingIndex": 0,
              },
            ],
            "id": "question-2",
            "isAnswer": false,
            "parentMessageId": "00000000-0000-0000-0000-000000000000",
            "siblingIndex": 0,
          },
          {
            "children": [
              {
                "children": [],
                "id": "4",
                "isAnswer": true,
                "parentMessageId": "question-4",
                "siblingIndex": 0,
              },
            ],
            "id": "question-4",
            "isAnswer": false,
            "parentMessageId": "1",
            "siblingIndex": 1,
          },
        ],
        "id": "1",
        "isAnswer": true,
        "parentMessageId": "question-1",
        "siblingIndex": 0,
      },
    ],
    "id": "question-1",
    "isAnswer": false,
    "parentMessageId": "00000000-0000-0000-0000-000000000000",
    "siblingIndex": 0,
  },
]
`;

exports[`buildChatItemTree should build a tree from a list of multi root nodes chat items 1`] = `
[
  {
    "children": [
      {
        "children": [
          {
            "children": [
              {
                "children": [
                  {
                    "children": [
                      {
                        "children": [],
                        "id": "3",
                        "isAnswer": true,
                        "parentMessageId": "question-3",
                        "siblingIndex": 0,
                      },
                    ],
                    "id": "question-3",
                    "isAnswer": false,
                    "parentMessageId": "2",
                    "siblingIndex": 0,
                  },
                ],
                "id": "2",
                "isAnswer": true,
                "parentMessageId": "question-2",
                "siblingIndex": 0,
              },
            ],
            "id": "question-2",
            "isAnswer": false,
            "parentMessageId": "1",
            "siblingIndex": 0,
          },
          {
            "children": [
              {
                "children": [],
                "id": "4",
                "isAnswer": true,
                "parentMessageId": "question-4",
                "siblingIndex": 0,
              },
            ],
            "id": "question-4",
            "isAnswer": false,
            "parentMessageId": "1",
            "siblingIndex": 1,
          },
        ],
        "id": "1",
        "isAnswer": true,
        "parentMessageId": "question-1",
        "siblingIndex": 0,
      },
    ],
    "id": "question-1",
    "isAnswer": false,
    "parentMessageId": null,
    "siblingIndex": 0,
  },
  {
    "children": [
      {
        "children": [],
        "id": "5",
        "isAnswer": true,
        "parentMessageId": "question-5",
        "siblingIndex": 0,
      },
    ],
    "id": "question-5",
    "isAnswer": false,
    "parentMessageId": null,
    "siblingIndex": 1,
  },
]
`;

exports[`buildChatItemTree should build a tree from a list of multi root nodes chat items with legacy chat items 1`] = `
[
  {
    "children": [
      {
        "children": [
          {
            "children": [
              {
                "children": [
                  {
                    "children": [
                      {
                        "children": [],
                        "id": "3",
                        "isAnswer": true,
                        "parentMessageId": "question-3",
                        "siblingIndex": 0,
                      },
                    ],
                    "id": "question-3",
                    "isAnswer": false,
                    "parentMessageId": "00000000-0000-0000-0000-000000000000",
                    "siblingIndex": 0,
                  },
                ],
                "id": "2",
                "isAnswer": true,
                "parentMessageId": "question-2",
                "siblingIndex": 0,
              },
            ],
            "id": "question-2",
            "isAnswer": false,
            "parentMessageId": "00000000-0000-0000-0000-000000000000",
            "siblingIndex": 0,
          },
          {
            "children": [
              {
                "children": [],
                "id": "4",
                "isAnswer": true,
                "parentMessageId": "question-4",
                "siblingIndex": 0,
              },
            ],
            "id": "question-4",
            "isAnswer": false,
            "parentMessageId": "1",
            "siblingIndex": 1,
          },
        ],
        "id": "1",
        "isAnswer": true,
        "parentMessageId": "question-1",
        "siblingIndex": 0,
      },
    ],
    "id": "question-1",
    "isAnswer": false,
    "parentMessageId": "00000000-0000-0000-0000-000000000000",
    "siblingIndex": 0,
  },
  {
    "children": [
      {
        "children": [],
        "id": "5",
        "isAnswer": true,
        "parentMessageId": "question-5",
        "siblingIndex": 0,
      },
    ],
    "id": "question-5",
    "isAnswer": false,
    "parentMessageId": null,
    "siblingIndex": 1,
  },
]
`;
